### PR TITLE
fix field sharing for sample/frame document cls

### PR DESCRIPTION
**Pull request by swheaton**
## What changes are proposed in this pull request?

It's not that big of a bug right now but it WAS confounding for an upcoming PR adding `created_at` to fields.
I say not that big because it only happens under specific circumstances and it's not common to save stuff to Field definitions.

Basically the issue is that when we create a subclass of `foo.DatasetSampleDocument` with `type()`, mongoengine seems to be populating `cls._fields` with a new list of _shallow_ copies of the original fields. That means any changes will be shared amongst some other datasets past or future. The devious part is if you call `field.save()` then it actually writes back to the field schema of `field._dataset` which may not be the dataset you intended.

New fields or loaded datasets (or default embedded fields such as `metadata`) are unaffected, because they will get new `Field` instances constructed for them.

## How is this patch tested? If it is not, please explain why.

From the added test...

```python
import fiftyone as fo

dataset1 = fo.Dataset()
dataset2 = fo.Dataset()

f = dataset1.get_field("filepath")
f.info = {"foo": "bar"}
f.save()

fo.Dataset._instances.clear()
dataset2b = fo.load_dataset(dataset2.name)
assert dataset2b.get_field("filepath") is None
```

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of field declarations to prevent unintended modifications to the original field object.

- **Tests**
	- Introduced a new test to verify the behavior of shared field metadata across different dataset instances, ensuring data integrity.

### CODE REVIEW

**Comment by swheaton** [comment]
Weird :bug: I just squashed you guys should check out 😃

File: fiftyone/core/dataset.py [hunk 1]
```diff
@@ -7617,9 +7617,10 @@ def _declare_fields(dataset, doc_cls, field_docs=None):
 
         if isinstance(field, fof.EmbeddedDocumentField):
             field = foo.create_field(field_name, **foo.get_field_kwargs(field))
-            doc_cls._declare_field(dataset, field_name, field)
         else:
-            field._set_dataset(dataset, field_name)
+            field = field.copy()
+            doc_cls._declare_field(dataset, field_name, field)
```
Comment (brimoor) [inline fiftyone/core/dataset.py:7622]: call `_declare_field()` twice?

Comment (swheaton) [reply fiftyone/core/dataset.py]: seconds too early ;)

**Review by brimoor — APPROVED** [review]
LGTM (and nice find 🧙‍♂️)

**Pull request merged**